resqpy/olio/base.py

```diff
@@ -0,0 +1,163 @@
+        self.originator = originator
+
+        if uuid is None:
+            self.uuid = bu.new_uuid()
+            logger.debug(f"Created new uuid for object {self}")
+        else:
+            self.uuid = uuid
+            logger.debug(f"Loading existing object {self}")
+            self.load_from_xml()
+    
+    # Define attributes self.part and self.root, using uuid as the primary key
+
+    @property
+    def part(self):
+        """Part corresponding to self.uuid"""
+
+        if self.uuid is None:
+            raise ValueError('Cannot get part if uuid is None')
+        return self.model.part_for_uuid(self.uuid)
+
+    @property
+    def root(self):
+        """Node corresponding to self.uuid"""
+
+        if self.uuid is None:
+            raise ValueError('Cannot get root if uuid is None')
+        return self.model.root_for_uuid(self.uuid)
+    
+    @root.setter
+    def root(self, value):
+        """Update self.uuid to match new root. Ensure root is added as a part"""
+
+        new_uuid = rqet.uuid_for_part_root(value)
+        if new_uuid is None:
+            raise ValueError("Cannot set uuid to be None")
+        self.uuid = new_uuid
+
+    def load_from_xml(self):
+        """Load citation block from XML"""
+
+        # Citation block
+        self.title = rqet.find_nested_tags_text(self.root, ['Citation', 'Title'])
+        self.originator = rqet.find_nested_tags_text(self.root, ['Citation', 'Originator'])
+
+    def create_xml(self, title=None, originator=None, ext_uuid=None, add_as_part=True):
+        """Write citation block to XML
+        
+        Args:
+            title (string): used as the citation Title text; should usually refer to the well name in a
+                human readable way
+            originator (string, optional): the name of the human being who created the deviation survey part;
+                default is to use the login name
+            add_as_part (boolean, default True): if True, the newly created xml node is added as a part
+                in the model
+
+        Returns:
+            node: the newly created root node
+        """
+
+        assert self.uuid is not None
+
+        if ext_uuid is None: ext_uuid = self.model.h5_uuid()
+
+        # Create the root node
+        node = self.model.new_obj_node(self._content_type)
+        node.attrib['uuid'] = str(self.uuid)
+
+        if add_as_part:
+            self.model.add_part(self._content_type, self.uuid, node)
+            assert self.root is not None   
+
+        # Citation block
+        if title: self.title = title
+        if originator: self.originator = originator
+        self.model.create_citation(
+            root=node, title=self.title, originator=self.originator
+        )
+        return node
+
+    # Generic magic methods
+
+    def __eq__(self, other):
+        """Implements equals operator. By default, compare objects using uuid"""
+        return self.uuid is not None and self.uuid == getattr(other, "uuid", None)
```

Comment:
resqpy.olio.uuid contains a function: matching_uuids() which is a safer way to compare uuids than ==

Reply:
Thanks, will do.

I was trying to understand the logic in some is_equivalent methods, maybe something more nuanced would be better?

```python
return isinstance(other, self.__class__) and any(
         self is other,
         bu.matching_uuids(self.uuid, other.uuid),
         all(
            self.title== getattr(other, "title", None),
            (equivalent_extra_metadata(self, other) or not check_extra_metadata))
      )
```

Or perhaps just comparing on uuids is simplest, given that is the point of them?